tests/integration.rs

```diff
@@ -19,6 +19,339 @@ use core::time::Duration;
+    );
+    let expired_demo_root_3: (&[u8], &str) = (
+        include_bytes!("amazon/expired.rootca3.demo.amazontrust.com.cer"),
+        "expired.rootca3.demo.amazontrust.com",
+    );
+    let expired_demo_root_4: (&[u8], &str) = (
+        include_bytes!("amazon/expired.rootca4.demo.amazontrust.com.cer"),
+        "expired.rootca4.demo.amazontrust.com",
+    );
+
+    for &(cert, dns_name) in &[
+        valid_demo_root_1,
+        valid_demo_root_2,
+        valid_demo_root_3,
+        valid_demo_root_4,
+        revoked_demo_root_1,
+        revoked_demo_root_2,
+        revoked_demo_root_3,
+        revoked_demo_root_4,
+        expired_demo_root_1,
+        expired_demo_root_2,
+        expired_demo_root_3,
+        expired_demo_root_4,
+    ] {
+        let cert = CertificateDer::from(cert);
+        let cert = webpki::EndEntityCert::try_from(&cert).unwrap();
+
+        assert_eq!(
+            Ok(()),
+            cert.verify_is_valid_for_subject_name(
+                &pki_types::ServerName::try_from(dns_name).unwrap()
+            ),
+        );
+    }
+
+    let time = UnixTime::since_unix_epoch(Duration::from_secs(1_740_304_936)); // Sun Feb 23 02:02:16 PST 2025
+
+    for &(cert, _dns_name) in &[
+        valid_demo_root_1,
+        valid_demo_root_2,
+        valid_demo_root_3,
+        valid_demo_root_4,
+    ] {
+        let cert = CertificateDer::from(cert);
+        let cert = webpki::EndEntityCert::try_from(&cert).unwrap();
+
+        for &crls in &[
+            None,
+            Some(&roots_crls),
+            Some(&intermediates_crls),
+            Some(&all_crls),
+        ] {
+            assert!(
+                cert.verify_for_usage(
+                    webpki::ALL_VERIFICATION_ALGS,
+                    &anchors,
+                    &intermediates,
+                    time,
+                    KeyUsage::server_auth(),
+                    crls.map(|l| revocation_options_for_test(l)),
+                    None,
+                )
+                .is_ok()
+            );
+
+            assert!(
+                cert.verify_for_usage(
+                    webpki::ALL_VERIFICATION_ALGS,
+                    &legacy_anchors,
+                    &intermediates_legacy,
+                    time,
+                    KeyUsage::server_auth(),
+                    crls.map(|l| revocation_options_for_test(l)),
+                    None,
+                )
+                .is_ok()
+            );
+
+            let path = cert
+                .verify_for_usage(
+                    webpki::ALL_VERIFICATION_ALGS,
+                    &all_anchors,
+                    &intermediates_legacy,
+                    time,
+                    KeyUsage::server_auth(),
+                    crls.map(|l| revocation_options_for_test(l)),
+                    None,
+                )
+                .unwrap();
+
+            // verify should find shortest path
+            assert!(anchors.contains(path.anchor()));
```

Comment:
@cpu @djc 
I’ve scanned the codebase and this is the only test usage of `verify_for_usage` where the computed path is checked, and specifically this check establishes that the shortest path is found. With the trust anchors and intermediates given here, there are two valid paths for every leaf cert tested.

Reply:
>  specifically this check establishes that the shortest path is found. With the trust anchors and intermediates given here, there are two valid paths for every leaf cert tested.

I believe the [better TLS](https://bettertls.com/) test coverage in [tests/better_tls.rs](https://github.com/rustls/webpki/blob/main/tests/better_tls.rs) has a variety of test cases that exercise path building like this. Did you take a peek at those?

Reply:
I didn’t, but now I’m looking
- https://github.com/Netflix/bettertls/tree/master/test-suites/pathbuilding

I’m actually not seeing a test case that matches what is seen in figures 1 & 2 in
- https://aws.amazon.com/blogs/security/acm-will-no-longer-cross-sign-certificates-with-starfield-class-2-starting-august-2024/

of course, there are a lot of complex/pathological in that bettertls.

Note the integration of this crate with better_tls has no revocation checking.

https://github.com/rustls/webpki/blob/84b1463bdbe6bf03a5fdd0a0e1c7c3de2e3af702/tests/better_tls.rs#L79-L86

Reply:
> Note the integration of this crate with better_tls also has no revocation checking.

Fair point 👍